sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -488,6 +489,16 @@ object SimplifyBinaryComparison
         // False with inequality
         case a GreaterThan b if canSimplifyComparison(a, b, notNullExpressions) => FalseLiteral
         case a LessThan b if canSimplifyComparison(a, b, notNullExpressions) => FalseLiteral
+
+        // One side is Literal
+        case (a: Predicate) EqualTo TrueLiteral => a
+        case TrueLiteral EqualTo (b: Predicate) => b
+        case (a: Predicate) EqualTo FalseLiteral => Not(a)
+        case FalseLiteral EqualTo (b: Predicate) => Not(b)
+        case (a: Predicate) EqualNullSafe TrueLiteral => And(a, IsNotNull(a))
```

Comment:
Hmm, doesn't it also mean the predicate `a` will be evaluated more than once?

Reply:
Thanks. I will change this to only for the case `!a.nullable`. I originally thought it is ok to evaluate twice in exchange of pushing down the filter. With the restriction of `!a.nullable`, the evaluation will be one time.

Reply:
Let's say if an expression is not non-deterministic, evaluating it twice could return different results.

Reply:
Yea, deterministics is an issue. Performance maybe too, although we usually do subexpression elimination, but it possibly adds a bit extra cost. I think best practice should be not duplicate the evaluation.

Reply:
Thank you all, updated